lib/core/src/Cardano/Wallet.hs

```diff
@@ -969,6 +985,119 @@ listTransactions ctx wid mStart mEnd order = db & \DBLayer{..} -> do
+    -> ExceptT ErrSignCert IO (Tx, TxMeta, UTCTime, [TxWitness])
+signCert ctx wid argGenChange pwd
+    (CoinSelection ins outs chgs) _poolId = db & \DBLayer{..} -> do
+    withRootKey @_ @s ctx wid pwd ErrSignCertWithRootKey $ \xprv -> do
+        mapExceptT atomically $ do
+            cp <- withExceptT ErrSignCertNoSuchWallet $ withNoSuchWallet wid $
+                readCheckpoint (PrimaryKey wid)
+            let (changeOuts, s') = flip runState (getState cp) $
+                    forM chgs $ \c -> do
+                        addr <- state (genChange argGenChange)
+                        return $ TxOut addr c
+            allShuffledOuts <- liftIO $ shuffle (outs ++ changeOuts)
+
+            --deriving chimeric account
+            let accPrv = deriveAccountPrivateKey @k pwd xprv minBound
+            let _addrPrv =
+                    deriveAddressPrivateKey pwd accPrv MutableAccount minBound
+
+            {--
+            let keyFrom = isOwned (getState cp) (xprv, pwd)
+            (tx, wit) <- either
+                (throwE . ErrSignCert)
+                pure
+                (mkCertificateTx tl keyFrom (poolId, addrPrv, pwd) ins allShuffledOuts)
+            --}
+            withExceptT ErrSignCertNoSuchWallet $
+                putCheckpoint (PrimaryKey wid) (updateState s' cp)
+            let ourCoins (TxOut addr (Coin val)) =
+                    if fst (isOurs addr s')
+                        then Just (fromIntegral val)
+                        else Nothing
+            let amtOuts =
+                    sum (mapMaybe ourCoins allShuffledOuts)
+            let amtInps = fromIntegral $
+                    sum (getCoin . coin . snd <$> ins)
+            let txSlot =
+                    (currentTip cp) ^. #slotId
+            let meta = TxMeta
+                    { status = Pending
+                    , direction = Outgoing
+                    , slotId = txSlot
+                    , blockHeight = (currentTip cp) ^. #blockHeight
+                    , amount = Quantity (amtInps - amtOuts)
+                    }
+            let time = slotStartTime
+                    (slotParams (blockchainParameters cp))
+                    txSlot
+            return (undefined, meta, time, undefined)
+  where
+    db = ctx ^. dbLayer @s @k
+    _tl = ctx ^. transactionLayer @t @k
+
+-- | Broadcast a (signed) transaction with certificate delegation to the network.
+submitCert
+    :: forall ctx s t k.
+        ( HasNetworkLayer t ctx
+        , HasDBLayer s k ctx
+        )
+    => ctx
+    -> WalletId
+    -> PoolId
+    -> (Tx, TxMeta, [TxWitness])
+    -> ExceptT ErrSubmitCert IO ()
+submitCert ctx wid poolId (tx, meta, wit) = db & \DBLayer{..} -> do
+    withExceptT ErrSubmitCertNetwork $ postTx nw (tx, wit)
+    mapExceptT atomically $ withExceptT ErrSubmitCertNoSuchWallet $ do
+        putTxHistory (PrimaryKey wid) [(tx, meta)]
+        let (TxMeta _ _ sl _ _) = meta
+        putDelegationCertificate (PrimaryKey wid) poolId sl
```

Comment:
writing both tx history and delegation certificate in DB

Reply:
:ok_hand: 

suggestion:

```hs
putDelegationCertificate (PrimaryKey wid) poolId (meta ^. #slotId)
```

why ? Avoiding pattern-maching here makes for code that is slightly easier to maintain. Adding fields to `TxMeta` won't require us to go back to this very point to add an extra `_`. Alternatively, using `NamedFieldPuns`, I'd also favor the following approach:

```hs
let TxMeta{sl} = meta
putDelegationCertificate (PrimaryKey wid) poolId sl
```

I'd still favor the first one because the entire line is then captured in a single expression and is easier to load in brain :)

Reply:
opted for lens version